lib/menubar/utils.lua

```diff
@@ -154,48 +156,63 @@ local icon_lookup_path = nil
 --- Get a list of icon lookup paths.
 -- @treturn table A list of directories, without trailing slash.
 local function get_icon_lookup_path()
-    if not icon_lookup_path then
-        local add_if_readable = function(t, path)
+    if icon_lookup_path then return icon_lookup_path end
+
+    local function ensure_args(t, paths)
+        if type(paths) == 'string' then paths = { paths } end
+        return t or {}, paths
+    end
+
+    local function add_if_readable(t, paths)
+        t, paths = ensure_args(t, paths)
+
+        for _, path in ipairs(paths) do
             if gfs.dir_readable(path) then
                 table.insert(t, path)
             end
         end
-        icon_lookup_path = {}
-        local icon_theme_paths = {}
-        local icon_theme = theme.icon_theme
-        local paths = glib.get_system_data_dirs()
-        table.insert(paths, 1, glib.get_user_data_dir())
-        table.insert(paths, 1, glib.build_filenamev({glib.get_home_dir(),
-                                                     '.icons'}))
-        for _,dir in ipairs(paths) do
-            local icons_dir = glib.build_filenamev({dir, 'icons'})
-            if gfs.dir_readable(icons_dir) then
-                if icon_theme then
-                    add_if_readable(icon_theme_paths,
-                                    glib.build_filenamev({icons_dir,
-                                                         icon_theme}))
-                end
-                -- Fallback theme.
-                add_if_readable(icon_theme_paths,
-                                glib.build_filenamev({icons_dir, 'hicolor'}))
-            end
-        end
-        for _, icon_theme_directory in ipairs(icon_theme_paths) do
-            for _, size in ipairs(all_icon_sizes) do
-                add_if_readable(icon_lookup_path,
-                                glib.build_filenamev({icon_theme_directory,
-                                                      size, 'apps'}))
-            end
+        return t
+    end
+
+    local function add_with_dir(t, paths, dir)
+        t, paths = ensure_args(t, paths)
+        dir = { nil, dir }
+
+        for _, path in ipairs(paths) do
+            dir[1] = path
+            table.insert(t, glib.build_filenamev(dir))
         end
-        for _,dir in ipairs(paths)do
-            -- lowest priority fallbacks
-            add_if_readable(icon_lookup_path,
-                            glib.build_filenamev({dir, 'pixmaps'}))
-            add_if_readable(icon_lookup_path,
-                            glib.build_filenamev({dir, 'icons'}))
+        return t
+    end
+
+    icon_lookup_path = {}
+    local theme_priority = { 'hicolor' }
+    if theme.icon_theme then table.insert(theme_priority, 1, theme.icon_theme) end
+
+    local paths = add_with_dir({}, glib.get_home_dir(), '.icons')
+    add_with_dir(paths, {
+        glib.get_user_data_dir(),           -- $XDG_DATA_HOME, typically $HOME/.local/share
```

Comment:
Previously `get_user_data_dir` was before `~/.icons`, now the order is swapped. Was this done on purpose? Why? (I'm too lazy to pull up the spec right now and I fear I won't get this question answered there anyway...)

Without your commit and with the following patch:
```diff
diff --git a/lib/menubar/utils.lua b/lib/menubar/utils.lua
index 49a6225b..a7c996b0 100644
--- a/lib/menubar/utils.lua
+++ b/lib/menubar/utils.lua
@@ -156,7 +156,7 @@ local icon_lookup_path = nil
 local function get_icon_lookup_path()
     if not icon_lookup_path then
         local add_if_readable = function(t, path)
-            if gfs.dir_readable(path) then
+            if true or gfs.dir_readable(path) then
                 table.insert(t, path)
             end
         end
@@ -169,7 +169,7 @@ local function get_icon_lookup_path()
                                                      '.icons'}))
         for _,dir in ipairs(paths) do
             local icons_dir = glib.build_filenamev({dir, 'icons'})
-            if gfs.dir_readable(icons_dir) then
+            if true or gfs.dir_readable(icons_dir) then
                 if icon_theme then
                     add_if_readable(icon_theme_paths,
                                     glib.build_filenamev({icons_dir,
@@ -195,6 +195,7 @@ local function get_icon_lookup_path()
                             glib.build_filenamev({dir, 'icons'}))
         end
     end
+    require("gears.debug").dump(icon_lookup_path)
     return icon_lookup_path
 end
 
```
I get this output (sigh, flatpack...):
```
table: 0x1c6d2c0
  1 : /dev/null/.icons/icons/hicolor/128x128/apps (string)
  2 : /dev/null/.icons/icons/hicolor/96x96/apps (string)
  3 : /dev/null/.icons/icons/hicolor/72x72/apps (string)
  4 : /dev/null/.icons/icons/hicolor/64x64/apps (string)
  5 : /dev/null/.icons/icons/hicolor/48x48/apps (string)
  6 : /dev/null/.icons/icons/hicolor/36x36/apps (string)
  7 : /dev/null/.icons/icons/hicolor/32x32/apps (string)
  8 : /dev/null/.icons/icons/hicolor/24x24/apps (string)
  9 : /dev/null/.icons/icons/hicolor/22x22/apps (string)
  10 : /dev/null/.icons/icons/hicolor/16x16/apps (string)
  11 : /dev/null/.local/share/icons/hicolor/128x128/apps (string)
  12 : /dev/null/.local/share/icons/hicolor/96x96/apps (string)
  13 : /dev/null/.local/share/icons/hicolor/72x72/apps (string)
  14 : /dev/null/.local/share/icons/hicolor/64x64/apps (string)
  15 : /dev/null/.local/share/icons/hicolor/48x48/apps (string)
  16 : /dev/null/.local/share/icons/hicolor/36x36/apps (string)
  17 : /dev/null/.local/share/icons/hicolor/32x32/apps (string)
  18 : /dev/null/.local/share/icons/hicolor/24x24/apps (string)
  19 : /dev/null/.local/share/icons/hicolor/22x22/apps (string)
  20 : /dev/null/.local/share/icons/hicolor/16x16/apps (string)
  21 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/128x128/apps (string)
  22 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/96x96/apps (string)
  23 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/72x72/apps (string)
  24 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/64x64/apps (string)
  25 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/48x48/apps (string)
  26 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/36x36/apps (string)
  27 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/32x32/apps (string)
  28 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/24x24/apps (string)
  29 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/22x22/apps (string)
  30 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/apps (string)
  31 : /var/lib/flatpak/exports/share/icons/hicolor/128x128/apps (string)
  32 : /var/lib/flatpak/exports/share/icons/hicolor/96x96/apps (string)
  33 : /var/lib/flatpak/exports/share/icons/hicolor/72x72/apps (string)
  34 : /var/lib/flatpak/exports/share/icons/hicolor/64x64/apps (string)
  35 : /var/lib/flatpak/exports/share/icons/hicolor/48x48/apps (string)
  36 : /var/lib/flatpak/exports/share/icons/hicolor/36x36/apps (string)
  37 : /var/lib/flatpak/exports/share/icons/hicolor/32x32/apps (string)
  38 : /var/lib/flatpak/exports/share/icons/hicolor/24x24/apps (string)
  39 : /var/lib/flatpak/exports/share/icons/hicolor/22x22/apps (string)
  40 : /var/lib/flatpak/exports/share/icons/hicolor/16x16/apps (string)
  41 : /usr/local/share/icons/hicolor/128x128/apps (string)
  42 : /usr/local/share/icons/hicolor/96x96/apps (string)
  43 : /usr/local/share/icons/hicolor/72x72/apps (string)
  44 : /usr/local/share/icons/hicolor/64x64/apps (string)
  45 : /usr/local/share/icons/hicolor/48x48/apps (string)
  46 : /usr/local/share/icons/hicolor/36x36/apps (string)
  47 : /usr/local/share/icons/hicolor/32x32/apps (string)
  48 : /usr/local/share/icons/hicolor/24x24/apps (string)
  49 : /usr/local/share/icons/hicolor/22x22/apps (string)
  50 : /usr/local/share/icons/hicolor/16x16/apps (string)
  51 : /usr/share/icons/hicolor/128x128/apps (string)
  52 : /usr/share/icons/hicolor/96x96/apps (string)
  53 : /usr/share/icons/hicolor/72x72/apps (string)
  54 : /usr/share/icons/hicolor/64x64/apps (string)
  55 : /usr/share/icons/hicolor/48x48/apps (string)
  56 : /usr/share/icons/hicolor/36x36/apps (string)
  57 : /usr/share/icons/hicolor/32x32/apps (string)
  58 : /usr/share/icons/hicolor/24x24/apps (string)
  59 : /usr/share/icons/hicolor/22x22/apps (string)
  60 : /usr/share/icons/hicolor/16x16/apps (string)
  61 : /dev/null/.icons/pixmaps (string)
  62 : /dev/null/.icons/icons (string)
  63 : /dev/null/.local/share/pixmaps (string)
  64 : /dev/null/.local/share/icons (string)
  65 : /home/psychon/.local/share/flatpak/exports/share/pixmaps (string)
  66 : /home/psychon/.local/share/flatpak/exports/share/icons (string)
  67 : /var/lib/flatpak/exports/share/pixmaps (string)
  68 : /var/lib/flatpak/exports/share/icons (string)
  69 : /usr/local/share/pixmaps (string)
  70 : /usr/local/share/icons (string)
  71 : /usr/share/pixmaps (string)
  72 : /usr/share/icons (string)
```
Adding the following patch ontop of your PR:
```diff
diff --git a/lib/menubar/utils.lua b/lib/menubar/utils.lua
index 83b86490..ab2503af 100644
--- a/lib/menubar/utils.lua
+++ b/lib/menubar/utils.lua
@@ -167,7 +167,7 @@ local function get_icon_lookup_path()
         t, paths = ensure_args(t, paths)
 
         for _, path in ipairs(paths) do
-            if gfs.dir_readable(path) then
+            if true or gfs.dir_readable(path) then
                 table.insert(t, path)
             end
         end
@@ -212,7 +212,9 @@ local function get_icon_lookup_path()
     end
     add_if_readable(icon_lookup_path, app_in_theme_paths)
 
-    return add_if_readable(icon_lookup_path, paths)
+    local ret = add_if_readable(icon_lookup_path, paths)
+    require("gears.debug").dump(ret)
+    return ret
 end
 
 --- Remove CR newline from the end of the string.
```
results in
```
table: 0x2570150
  1 : /dev/null/.icons/hicolor/scalable/apps (string)
  2 : /dev/null/.icons/hicolor/128x128/apps (string)
  3 : /dev/null/.icons/hicolor/96x96/apps (string)
  4 : /dev/null/.icons/hicolor/72x72/apps (string)
  5 : /dev/null/.icons/hicolor/64x64/apps (string)
  6 : /dev/null/.icons/hicolor/48x48/apps (string)
  7 : /dev/null/.icons/hicolor/36x36/apps (string)
  8 : /dev/null/.icons/hicolor/32x32/apps (string)
  9 : /dev/null/.icons/hicolor/24x24/apps (string)
  10 : /dev/null/.icons/hicolor/22x22/apps (string)
  11 : /dev/null/.icons/hicolor/16x16/apps (string)
  12 : /dev/null/.local/share/icons/hicolor/scalable/apps (string)
  13 : /dev/null/.local/share/icons/hicolor/128x128/apps (string)
  14 : /dev/null/.local/share/icons/hicolor/96x96/apps (string)
  15 : /dev/null/.local/share/icons/hicolor/72x72/apps (string)
  16 : /dev/null/.local/share/icons/hicolor/64x64/apps (string)
  17 : /dev/null/.local/share/icons/hicolor/48x48/apps (string)
  18 : /dev/null/.local/share/icons/hicolor/36x36/apps (string)
  19 : /dev/null/.local/share/icons/hicolor/32x32/apps (string)
  20 : /dev/null/.local/share/icons/hicolor/24x24/apps (string)
  21 : /dev/null/.local/share/icons/hicolor/22x22/apps (string)
  22 : /dev/null/.local/share/icons/hicolor/16x16/apps (string)
  23 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/scalable/apps (string)
  24 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/128x128/apps (string)
  25 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/96x96/apps (string)
  26 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/72x72/apps (string)
  27 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/64x64/apps (string)
  28 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/48x48/apps (string)
  29 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/36x36/apps (string)
  30 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/32x32/apps (string)
  31 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/24x24/apps (string)
  32 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/22x22/apps (string)
  33 : /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/apps (string)
  34 : /var/lib/flatpak/exports/share/icons/hicolor/scalable/apps (string)
  35 : /var/lib/flatpak/exports/share/icons/hicolor/128x128/apps (string)
  36 : /var/lib/flatpak/exports/share/icons/hicolor/96x96/apps (string)
  37 : /var/lib/flatpak/exports/share/icons/hicolor/72x72/apps (string)
  38 : /var/lib/flatpak/exports/share/icons/hicolor/64x64/apps (string)
  39 : /var/lib/flatpak/exports/share/icons/hicolor/48x48/apps (string)
  40 : /var/lib/flatpak/exports/share/icons/hicolor/36x36/apps (string)
  41 : /var/lib/flatpak/exports/share/icons/hicolor/32x32/apps (string)
  42 : /var/lib/flatpak/exports/share/icons/hicolor/24x24/apps (string)
  43 : /var/lib/flatpak/exports/share/icons/hicolor/22x22/apps (string)
  44 : /var/lib/flatpak/exports/share/icons/hicolor/16x16/apps (string)
  45 : /usr/local/share/icons/hicolor/scalable/apps (string)
  46 : /usr/local/share/icons/hicolor/128x128/apps (string)
  47 : /usr/local/share/icons/hicolor/96x96/apps (string)
  48 : /usr/local/share/icons/hicolor/72x72/apps (string)
  49 : /usr/local/share/icons/hicolor/64x64/apps (string)
  50 : /usr/local/share/icons/hicolor/48x48/apps (string)
  51 : /usr/local/share/icons/hicolor/36x36/apps (string)
  52 : /usr/local/share/icons/hicolor/32x32/apps (string)
  53 : /usr/local/share/icons/hicolor/24x24/apps (string)
  54 : /usr/local/share/icons/hicolor/22x22/apps (string)
  55 : /usr/local/share/icons/hicolor/16x16/apps (string)
  56 : /usr/share/icons/hicolor/scalable/apps (string)
  57 : /usr/share/icons/hicolor/128x128/apps (string)
  58 : /usr/share/icons/hicolor/96x96/apps (string)
  59 : /usr/share/icons/hicolor/72x72/apps (string)
  60 : /usr/share/icons/hicolor/64x64/apps (string)
  61 : /usr/share/icons/hicolor/48x48/apps (string)
  62 : /usr/share/icons/hicolor/36x36/apps (string)
  63 : /usr/share/icons/hicolor/32x32/apps (string)
  64 : /usr/share/icons/hicolor/24x24/apps (string)
  65 : /usr/share/icons/hicolor/22x22/apps (string)
  66 : /usr/share/icons/hicolor/16x16/apps (string)
  67 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/scalable/apps (string)
  68 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/128x128/apps (string)
  69 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/96x96/apps (string)
  70 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/72x72/apps (string)
  71 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/64x64/apps (string)
  72 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/48x48/apps (string)
  73 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/36x36/apps (string)
  74 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/32x32/apps (string)
  75 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/24x24/apps (string)
  76 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/22x22/apps (string)
  77 : /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/16x16/apps (string)
  78 : /var/lib/flatpak/exports/share/pixmaps/hicolor/scalable/apps (string)
  79 : /var/lib/flatpak/exports/share/pixmaps/hicolor/128x128/apps (string)
  80 : /var/lib/flatpak/exports/share/pixmaps/hicolor/96x96/apps (string)
  81 : /var/lib/flatpak/exports/share/pixmaps/hicolor/72x72/apps (string)
  82 : /var/lib/flatpak/exports/share/pixmaps/hicolor/64x64/apps (string)
  83 : /var/lib/flatpak/exports/share/pixmaps/hicolor/48x48/apps (string)
  84 : /var/lib/flatpak/exports/share/pixmaps/hicolor/36x36/apps (string)
  85 : /var/lib/flatpak/exports/share/pixmaps/hicolor/32x32/apps (string)
  86 : /var/lib/flatpak/exports/share/pixmaps/hicolor/24x24/apps (string)
  87 : /var/lib/flatpak/exports/share/pixmaps/hicolor/22x22/apps (string)
  88 : /var/lib/flatpak/exports/share/pixmaps/hicolor/16x16/apps (string)
  89 : /usr/local/share/pixmaps/hicolor/scalable/apps (string)
  90 : /usr/local/share/pixmaps/hicolor/128x128/apps (string)
  91 : /usr/local/share/pixmaps/hicolor/96x96/apps (string)
  92 : /usr/local/share/pixmaps/hicolor/72x72/apps (string)
  93 : /usr/local/share/pixmaps/hicolor/64x64/apps (string)
  94 : /usr/local/share/pixmaps/hicolor/48x48/apps (string)
  95 : /usr/local/share/pixmaps/hicolor/36x36/apps (string)
  96 : /usr/local/share/pixmaps/hicolor/32x32/apps (string)
  97 : /usr/local/share/pixmaps/hicolor/24x24/apps (string)
  98 : /usr/local/share/pixmaps/hicolor/22x22/apps (string)
  99 : /usr/local/share/pixmaps/hicolor/16x16/apps (string)
  100 : /usr/share/pixmaps/hicolor/scalable/apps (string)
  101 : /usr/share/pixmaps/hicolor/128x128/apps (string)
  102 : /usr/share/pixmaps/hicolor/96x96/apps (string)
  103 : /usr/share/pixmaps/hicolor/72x72/apps (string)
  104 : /usr/share/pixmaps/hicolor/64x64/apps (string)
  105 : /usr/share/pixmaps/hicolor/48x48/apps (string)
  106 : /usr/share/pixmaps/hicolor/36x36/apps (string)
  107 : /usr/share/pixmaps/hicolor/32x32/apps (string)
  108 : /usr/share/pixmaps/hicolor/24x24/apps (string)
  109 : /usr/share/pixmaps/hicolor/22x22/apps (string)
  110 : /usr/share/pixmaps/hicolor/16x16/apps (string)
  111 : /dev/null/.icons (string)
  112 : /dev/null/.local/share/icons (string)
  113 : /home/psychon/.local/share/flatpak/exports/share/icons (string)
  114 : /var/lib/flatpak/exports/share/icons (string)
  115 : /usr/local/share/icons (string)
  116 : /usr/share/icons (string)
  117 : /home/psychon/.local/share/flatpak/exports/share/pixmaps (string)
  118 : /var/lib/flatpak/exports/share/pixmaps (string)
  119 : /usr/local/share/pixmaps (string)
  120 : /usr/share/pixmaps (string)
```
So this patch increases the potential size of the search path by factor almost two. I'm looking forward to the resulting bug reports about slowness. ;-)
The diff between the two search paths is:
```diff
--- /tmp/before	2018-01-16 09:22:42.724480398 +0100
+++ /tmp/after	2018-01-16 09:22:42.807477999 +0100
@@ -1,13 +1,15 @@
- /dev/null/.icons/icons/hicolor/128x128/apps (string)
- /dev/null/.icons/icons/hicolor/96x96/apps (string)
- /dev/null/.icons/icons/hicolor/72x72/apps (string)
- /dev/null/.icons/icons/hicolor/64x64/apps (string)
- /dev/null/.icons/icons/hicolor/48x48/apps (string)
- /dev/null/.icons/icons/hicolor/36x36/apps (string)
- /dev/null/.icons/icons/hicolor/32x32/apps (string)
- /dev/null/.icons/icons/hicolor/24x24/apps (string)
- /dev/null/.icons/icons/hicolor/22x22/apps (string)
- /dev/null/.icons/icons/hicolor/16x16/apps (string)
+ /dev/null/.icons/hicolor/scalable/apps (string)
+ /dev/null/.icons/hicolor/128x128/apps (string)
+ /dev/null/.icons/hicolor/96x96/apps (string)
+ /dev/null/.icons/hicolor/72x72/apps (string)
+ /dev/null/.icons/hicolor/64x64/apps (string)
+ /dev/null/.icons/hicolor/48x48/apps (string)
+ /dev/null/.icons/hicolor/36x36/apps (string)
+ /dev/null/.icons/hicolor/32x32/apps (string)
+ /dev/null/.icons/hicolor/24x24/apps (string)
+ /dev/null/.icons/hicolor/22x22/apps (string)
+ /dev/null/.icons/hicolor/16x16/apps (string)
+ /dev/null/.local/share/icons/hicolor/scalable/apps (string)
  /dev/null/.local/share/icons/hicolor/128x128/apps (string)
  /dev/null/.local/share/icons/hicolor/96x96/apps (string)
  /dev/null/.local/share/icons/hicolor/72x72/apps (string)
@@ -18,6 +20,7 @@
  /dev/null/.local/share/icons/hicolor/24x24/apps (string)
  /dev/null/.local/share/icons/hicolor/22x22/apps (string)
  /dev/null/.local/share/icons/hicolor/16x16/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/scalable/apps (string)
  /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/128x128/apps (string)
  /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/96x96/apps (string)
  /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/72x72/apps (string)
@@ -28,6 +31,7 @@
  /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/24x24/apps (string)
  /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/22x22/apps (string)
  /home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/apps (string)
+ /var/lib/flatpak/exports/share/icons/hicolor/scalable/apps (string)
  /var/lib/flatpak/exports/share/icons/hicolor/128x128/apps (string)
  /var/lib/flatpak/exports/share/icons/hicolor/96x96/apps (string)
  /var/lib/flatpak/exports/share/icons/hicolor/72x72/apps (string)
@@ -38,6 +42,7 @@
  /var/lib/flatpak/exports/share/icons/hicolor/24x24/apps (string)
  /var/lib/flatpak/exports/share/icons/hicolor/22x22/apps (string)
  /var/lib/flatpak/exports/share/icons/hicolor/16x16/apps (string)
+ /usr/local/share/icons/hicolor/scalable/apps (string)
  /usr/local/share/icons/hicolor/128x128/apps (string)
  /usr/local/share/icons/hicolor/96x96/apps (string)
  /usr/local/share/icons/hicolor/72x72/apps (string)
@@ -48,6 +53,7 @@
  /usr/local/share/icons/hicolor/24x24/apps (string)
  /usr/local/share/icons/hicolor/22x22/apps (string)
  /usr/local/share/icons/hicolor/16x16/apps (string)
+ /usr/share/icons/hicolor/scalable/apps (string)
  /usr/share/icons/hicolor/128x128/apps (string)
  /usr/share/icons/hicolor/96x96/apps (string)
  /usr/share/icons/hicolor/72x72/apps (string)
@@ -58,15 +64,57 @@
  /usr/share/icons/hicolor/24x24/apps (string)
  /usr/share/icons/hicolor/22x22/apps (string)
  /usr/share/icons/hicolor/16x16/apps (string)
- /dev/null/.icons/pixmaps (string)
- /dev/null/.icons/icons (string)
- /dev/null/.local/share/pixmaps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/scalable/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/128x128/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/96x96/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/72x72/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/64x64/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/48x48/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/36x36/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/32x32/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/24x24/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/22x22/apps (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps/hicolor/16x16/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/scalable/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/128x128/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/96x96/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/72x72/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/64x64/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/48x48/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/36x36/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/32x32/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/24x24/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/22x22/apps (string)
+ /var/lib/flatpak/exports/share/pixmaps/hicolor/16x16/apps (string)
+ /usr/local/share/pixmaps/hicolor/scalable/apps (string)
+ /usr/local/share/pixmaps/hicolor/128x128/apps (string)
+ /usr/local/share/pixmaps/hicolor/96x96/apps (string)
+ /usr/local/share/pixmaps/hicolor/72x72/apps (string)
+ /usr/local/share/pixmaps/hicolor/64x64/apps (string)
+ /usr/local/share/pixmaps/hicolor/48x48/apps (string)
+ /usr/local/share/pixmaps/hicolor/36x36/apps (string)
+ /usr/local/share/pixmaps/hicolor/32x32/apps (string)
+ /usr/local/share/pixmaps/hicolor/24x24/apps (string)
+ /usr/local/share/pixmaps/hicolor/22x22/apps (string)
+ /usr/local/share/pixmaps/hicolor/16x16/apps (string)
+ /usr/share/pixmaps/hicolor/scalable/apps (string)
+ /usr/share/pixmaps/hicolor/128x128/apps (string)
+ /usr/share/pixmaps/hicolor/96x96/apps (string)
+ /usr/share/pixmaps/hicolor/72x72/apps (string)
+ /usr/share/pixmaps/hicolor/64x64/apps (string)
+ /usr/share/pixmaps/hicolor/48x48/apps (string)
+ /usr/share/pixmaps/hicolor/36x36/apps (string)
+ /usr/share/pixmaps/hicolor/32x32/apps (string)
+ /usr/share/pixmaps/hicolor/24x24/apps (string)
+ /usr/share/pixmaps/hicolor/22x22/apps (string)
+ /usr/share/pixmaps/hicolor/16x16/apps (string)
+ /dev/null/.icons (string)
  /dev/null/.local/share/icons (string)
- /home/psychon/.local/share/flatpak/exports/share/pixmaps (string)
  /home/psychon/.local/share/flatpak/exports/share/icons (string)
- /var/lib/flatpak/exports/share/pixmaps (string)
  /var/lib/flatpak/exports/share/icons (string)
- /usr/local/share/pixmaps (string)
  /usr/local/share/icons (string)
- /usr/share/pixmaps (string)
  /usr/share/icons (string)
+ /home/psychon/.local/share/flatpak/exports/share/pixmaps (string)
+ /var/lib/flatpak/exports/share/pixmaps (string)
+ /usr/local/share/pixmaps (string)
+ /usr/share/pixmaps (string)
```
Changes that I see are:
- `.icons` instead of `.icons/icons`: Good, that's part of what this wants to fix.
- The addition of `scalable`: Well, okay...
- `/dev/null/.icons/pixmaps`, `/dev/null/.local/share/pixmaps`, `/dev/null/.local/share/pixmaps`  are removed: Uhm, okay, these paths might be wrong anyway. I'm not sure.
- `/home/psychon/.local/share/flatpak/exports/share/pixmaps` is now interpreted as a directory containing themes while previously it directly contained icons. Since it has `pixmaps` in its name, this seems wrong...? The same change for all that other (geh, so many?!?) `/pixmaps/` directory definitely feels wrong to me. AFAIK it should just be e.g. `/usr/share/pixmaps` (well, this path exists later in the search path, but why are there these subdirectories at all?).

Reply:
I have no idea how flatpack managed to get included here, but yeah, it is some kind of sandbox utility.

Reply:
> I love it when people quote the spec on me. That should happen more often. :-)

Hey, happy to oblige :3

Ideally, the index.theme file should be parsed, it declares all the subdirectories so hammering the filesystem iterating over sizes wouldn't be needed, but if you are planning on using the gtk icon caches it would be redundant.

What are you using for benchmarking in lua?

Reply:
You mean something like `lib/menubar/index_theme.lua` which is part of the code that was deprecated because it is too **** slow? Yeah, that will surely help. ;-)

The problem is e.g. this:
```
$ grep Directories /usr/share/icons/Adwaita/index.theme 
Directories=8x8/emblems,16x16/actions,16x16/apps,16x16/categories,16x16/devices,16x16/emblems,16x16/emotes,16x16/mimetypes,16x16/places,16x16/status,22x22/actions,22x22/apps,22x22/categories,22x22/devices,22x22/emblems,22x22/emotes,22x22/mimetypes,22x22/places,22x22/status,24x24/actions,24x24/apps,24x24/categories,24x24/devices,24x24/emblems,24x24/emotes,24x24/mimetypes,24x24/places,24x24/status,32x32/actions,32x32/apps,32x32/categories,32x32/devices,32x32/emblems,32x32/emotes,32x32/mimetypes,32x32/places,32x32/status,48x48/actions,48x48/apps,48x48/categories,48x48/devices,48x48/emblems,48x48/emotes,48x48/mimetypes,48x48/places,48x48/status,64x64/actions,64x64/apps,64x64/categories,64x64/devices,64x64/emblems,64x64/emotes,64x64/mimetypes,64x64/places,64x64/status,96x96/actions,96x96/apps,96x96/categories,96x96/devices,96x96/emblems,96x96/emotes,96x96/mimetypes,96x96/places,96x96/status,256x256/actions,256x256/apps,256x256/categories,256x256/devices,256x256/emblems,256x256/emotes,256x256/mimetypes,256x256/places,256x256/status,512x512/actions,512x512/apps,512x512/categories,512x512/devices,512x512/emblems,512x512/mimetypes,512x512/places,512x512/status,scalable/actions,scalable/apps,scalable/categories,scalable/devices,scalable/emblems,scalable/emotes,scalable/mimetypes,scalable/places,scalable/status,scalable-up-to-32/status,
```
So the code based on `index.theme` would need to hammer the filesystem even more than the code that this PR touches.

I think that all of this needs to be rewritten in a way that (a) follows the spec (I think the deprecated code actually tried to do this) and (b) uses the cache file so that it is not awfully slow. Sadly, last time I looked at the existing code, I decided that incorporating the cache file in there does not really work.

> but if you are planning on using the gtk icon caches it would be redundant.

Problem is: The code still needs to work if the cache file is not present or outdated.... And preferably in a non-awfully-slow way.

> What are you using for benchmarking in lua?

Uhm... my first reaction was "nothing", but it was actually me that wrote `tests/test-benchmark.lua`. That code tries to find a number `n` so that `n` iterations of the target function takes about a second, and then does a division to compute the time per iteration. This _might_ actually work for LuaJIT in that the JIT already got a chance to optimise the code.

Is this what you had in mind?

(`test-benchmark.lua` happened because people claimed that tag switched are slow and proposed random patches that they felt would improve the situation, but no numbers were provided)

Reply:
I haven't looked that hard at `index_theme` below its interface so I don't know why it's so slow, but

> would need to hammer the filesystem even more

Why? For the purposes of what the path resolution does here (only finding `app/` icons), what it does now is
```
for icondir in icondirs
    if fs.dir_exists(icondir/theme_name)
        for size in icon_sizes
            if fs.dir_exists(icondir/theme_name/size/ .. 'apps')
                add_to_path
```
so for a theme, it hits the fs 1 + sizes times, compared to
```
for icondir in icondirs
    for line in fs.open(icondir/theme_name/ .. 'theme.index')
        if line:match('^Directories=')
             for dir in line:split(',')
                 if dir:match('/apps$')
                     add_to_path
```
which hits it once. Am I missing something here?

(To support 'icon theme inheritance' theme.index must also be parsed, since parent themes are declared there. I know that themes can also be extended by dropping icons in i.e. `~/.icons/hicolor` without requiring a `~/.icons/hicolor/index.theme`, but I consider resolving those a 'fallback case')

Furthermore, from my understanding, when giving `lookup_icon_uncached` a name like `firefox`, it traverses the path, asking fs for `firefox.{png,xpm,svg}`. So if a user has `/usr/share/icons/kewl_theme/<all-sizes>/apps` and `firefox.svg` is in `hicolor`, it hits the fs `sizes * 3` times in `kewl_theme` alone before finding it.

My immediate thought is something like
```
for icondir in icondirs
    for line in fs.open(icondir/theme_name/ .. 'theme.index')
        if line:match('^Directories=')
             for dir in line:split(',')
                 if dir:match('/apps$')
                     icon_index.add(icondir/theme_name/dir, fs.glob('*.svg *.png *.xpm'))
```
and resolve 'best icon file' from `icon_index` over traversing the path hammering the fs for every uncached icon.

> benchmarking

I was thinking some module/luarock along the lines of https://ruby-doc.org/stdlib-2.5.0/libdoc/benchmark/rdoc/Benchmark.html but if there's no 'defacto lua standard' I can just use/cook something from `test-benchmark`

Reply:
> Why? 

Mostly because of "what happens when no icon is found".
```diff
diff --git a/lib/menubar/icon_theme.lua b/lib/menubar/icon_theme.lua
index 358f0073..6500bd87 100644
--- a/lib/menubar/icon_theme.lua
+++ b/lib/menubar/icon_theme.lua
@@ -21,6 +21,11 @@ local string = string
 local table = table
 local math = math
 
+local function file_readable(path)
+    print(path)
+    return gfs.file_readable(path)
+end
+
 local get_pragmatic_base_directories = function()
     local dirs = {}
 
@@ -65,7 +70,7 @@ local get_default_icon_theme_name = function()
     for _, dir in ipairs(get_pragmatic_base_directories()) do
         for _, icon_theme_name in ipairs(icon_theme_names) do
             local filename = string.format("%s/%s/index.theme", dir, icon_theme_name)
-            if gfs.file_readable(filename) then
+            if file_readable(filename) then
                 return icon_theme_name
             end
         end
@@ -153,7 +158,7 @@ local lookup_icon = function(self, icon_name, icon_size)
                     local filename = string.format("%s/%s/%s/%s.%s",
                                                    basedir, self.icon_theme_name, subdir,
                                                    icon_name, ext)
-                    if gfs.file_readable(filename) then
+                    if file_readable(filename) then
                         return filename
                     else
                         checked_already[filename] = true
@@ -174,7 +179,7 @@ local lookup_icon = function(self, icon_name, icon_size)
                                                    basedir, self.icon_theme_name, subdir,
                                                    icon_name, ext)
                     if not checked_already[filename] then
-                        if gfs.file_readable(filename) then
+                        if file_readable(filename) then
                             closest_filename = filename
                             minimal_size = dist
                         end
@@ -210,7 +215,7 @@ local lookup_fallback_icon = function(self, icon_name)
             local filename = string.format("%s/%s.%s",
                                            dir,
                                            icon_name, ext)
-            if gfs.file_readable(filename) then
+            if file_readable(filename) then
                 return filename
             end
         end
```
makes `print("result:", require("menubar.icon_theme")():find_icon_path("firefox"))` result in:
```
/home/psychon/.local/share/flatpak/exports/share/icons/Adwaita/index.theme
/home/psychon/.local/share/flatpak/exports/share/icons/gnome/index.theme
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/index.theme
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/actions/firefox.png
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/actions/firefox.svg
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/actions/firefox.xpm
/usr/share/icons/hicolor/16x16/actions/firefox.png
/usr/share/icons/hicolor/16x16/actions/firefox.svg
/usr/share/icons/hicolor/16x16/actions/firefox.xpm
/usr/share/pixmaps/hicolor/16x16/actions/firefox.png
/usr/share/pixmaps/hicolor/16x16/actions/firefox.svg
/usr/share/pixmaps/hicolor/16x16/actions/firefox.xpm
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/animations/firefox.png
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/animations/firefox.svg
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/animations/firefox.xpm
/usr/share/icons/hicolor/16x16/animations/firefox.png
/usr/share/icons/hicolor/16x16/animations/firefox.svg
/usr/share/icons/hicolor/16x16/animations/firefox.xpm
/usr/share/pixmaps/hicolor/16x16/animations/firefox.png
/usr/share/pixmaps/hicolor/16x16/animations/firefox.svg
/usr/share/pixmaps/hicolor/16x16/animations/firefox.xpm
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/apps/firefox.png
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/apps/firefox.svg
/home/psychon/.local/share/flatpak/exports/share/icons/hicolor/16x16/apps/firefox.xpm
/usr/share/icons/hicolor/16x16/apps/firefox.png
result:	/usr/share/icons/hicolor/16x16/apps/firefox.png
```
That's not too bad. However, when looking up `DoesNotExist`, there are 3045 paths that are checked. For space reason I put the list in a gist: https://gist.github.com/psychon/3289ba38626b7026aa1da3f26bc2c75c
And AFAIK, none of those paths can really be left out (well, ignore the flatpak stuff). So, with a HDD, the seek time for all of those disk hits becomes quite noticable.

This code was just used to look up six icons in the menubar (category icons). Turns out that most people do not have those icons and just skipping their lookup with this code makes everything go fast.

Reply:
Sorry about the late reply, I'm sort of swamped atm. I was hoping I'd have time to write some actual code and have a good look at `icon_theme` before replying but ..

Doesn't this just show that `icon_theme` has the same problem as `lookup_icon_uncached`? Instead of running over the path once and indexing every icon file, it only indexes the paths, and then asks the fs for the file for potentially every path in icon_path.

I want to have a go at this when I have the time, but when that will be .. I currently don't know.